sdks/python/apache_beam/runners/common.py

```diff
@@ -648,24 +856,26 @@ def invoke_process(self,
         raise ValueError(
             'A RestrictionTracker %r was provided but DoFn does not have a '
             'RestrictionTrackerParam defined' % restriction_tracker)
-      from apache_beam.io import iobase
-      self.threadsafe_restriction_tracker = iobase.ThreadsafeRestrictionTracker(
-          restriction_tracker)
+      self.threadsafe_restriction_tracker = _ThreadsafeRestrictionTracker(
+          restriction_tracker, self._synchronized_lock)
       additional_kwargs[restriction_tracker_param] = (
-          iobase.RestrictionTrackerView(self.threadsafe_restriction_tracker))
-
-      if self.watermark_estimator:
-        # The watermark estimator needs to be reset for every element.
-        self.watermark_estimator.reset()
```

Comment:
Shouldn't we be setting it with the watermark estimator state that belongs to this element?

Reply:
With latest changes, the watermark_estimator has been created with estimator_state.